### PR TITLE
implement Sabre IMipHandler

### PR DESCRIPTION
**Pull request by georgehrke**
please test

### CODE REVIEW

**Comment by tanghus** [comment]
@georgehrke does this relate to an open issue? I don't even know what IMipHandler does ;)

**Comment by georgehrke** [comment]
https://github.com/owncloud/calendar/issues/318

**Comment by tanghus** [comment]
@joergmschulz @jancborchardt @re-fael @schmunk42 Can you test this patch? It's not a feature I ever use, so don't really know how to test.

**Comment by joergmschulz** [comment]
sorry, I cannot test currently - my OC server is down; sparepart is in mail... I test as soon as possible

**Comment by foux** [comment]
I just tested it, with the exact same error than before (even after restarting my nginx)

**Comment by georgehrke** [comment]
@foux What exact error msg do you get?

**Comment by foux** [comment]
Still the 403 error to CalDAVScheduleEventQueueableOperation

**Comment by georgehrke** [comment]
what client program?

**Comment by foux** [comment]
OS X Calendar on 10.9.2 (Mavericks)

**Comment by joergmschulz** [comment]
now I have my OC back - adding a debug message, I found out that the EMail address read by
$email = \OCP\Config::getAppValue('settings', 'email');
remains empty 
How can this happen?

**Comment by georgehrke** [comment]
sorry, getAppValue was wrong, can you check the new patch please

**Comment by joergmschulz** [comment]
sorry, \OCP\User::getUser(); returns an empty value

**Comment by joergmschulz** [comment]
if I set a fixed value for the email address when $email is empty, I get
{"app":"PHP","message":"Class 'Sabre\CalDAV\Plugin' not found at \/var\/www\/owncloud\/apps\/calendar\/appinfo\/remote.php#56","level":3,"time":"2014-03-14T20:00:31+00:00"}

so: two errors: The user won't be identified, Sabre plugin not found.
Hope this helps...

**Comment by georgehrke** [comment]
thanks, I'll look into it tomorrow.

**Comment by joergmschulz** [comment]
appreciated! Tell me if I can contribute with my limited resources and knowledge…

Am 14.03.2014 um 21:05 schrieb Georg Ehrke notifications@github.com:

**Comment by schmunk42** [comment]
So, I am kinda confused how to test the patch.

My calendar is installed via oC, but it's not a git repo. 
The Web-UI says **Calendar 0.6.3**, but the git repo here has a **6.0.2** tag?!

How should I test it, download and replace the calendar folder in `/apps`?

**Comment by georgehrke** [comment]
> The Web-UI says Calendar 0.6.3, but the git repo here has a 6.0.2 tag?!

the calendar is being shipped together with ownCloud core. That's why the tags are equal to the ownCloud version. 

> How should I test it, download and replace the calendar folder in /apps?

Download the patch file and apply it to the calendar app.
https://github.com/owncloud/calendar/pull/349/files

**Comment by schmunk42** [comment]
I applied the patch, but without luck.

I found this one: https://groups.google.com/d/msg/sabredav-discuss/BkvtenYWIWw/KCqhK8kaJtQJ
Seems to be fixed in SabreDAV 1.7 - which version does oC use, I bet you have a newer one already.

How can I test my e-mail functionality from oC? I have a LDAP login, so there's no send password functionality?

**Comment by brianjmurrell** [comment]
So I applied the patch in https://github.com/owncloud/calendar/pull/349/files on top of 6.0.2 but I still don't get meeting invites for meetings I create with participants.  I did make sure to fill out my e-mail address in my user profile.

Anything I can do to help debug?

Update: I did notice this morning in my apache error.log:

```
PHP Fatal error:  Class 'Sabre\\CalDAV\\Plugin' not found in /usr/src/owncloud/apps/calendar/appinfo/remote.php on line 49
```

That "backslash" syntax does look rather odd.

**Comment by brianjmurrell** [comment]
Hrm.  No advice here?  I am most willing to help move this along and am a software developer but have not really done any PHP thus far so the syntax is a bit strange to me still.

**Comment by joergmschulz** [comment]
Hello, 
I changed the code a bit and entered:

$email = \OCP\Config::getUserValue(OCP\USER::getUser(), 'settings', 'email');
OCP\Util::writeLog('calendar','remote.php user: '. $userName .' email: '.$email, OCP\Util::ERROR);

if($email == null) {
        $email = 'myEmailAddress@myDomain.local';
 }
if($email !== null) {
        $caldavPlugin = new Sabre_CalDAV_Plugin();
        $caldavPlugin->setIMipHandler(
                new Sabre_CalDAV_Schedule_IMip($email)
        );
        $server->addPlugin($caldavPlugin);
}

thus, you can get logging info.
As you see, the issue lies in the caldav lib somewhere.

Got stuck here….
Best regards
Jörg
Am 11.04.2014 um 00:20 schrieb Brian J. Murrell notifications@github.com:

**Comment by brianjmurrell** [comment]
@joergmschulz: Yeah, I got that far as well but am now getting:

```
{"app":"PHP","message":"Class 'Sabre\\CalDAV\\Plugin' not found at \/usr\/src\/owncloud\/apps\/calendar\/appinfo\/remote.php#55","level":3,"time":"2014-04-11T10:39:22+00:00"}
```

in the log.  Not sure why yet.

**Comment by joergmschulz** [comment]
Hello,
just compare the content of the patch with what I inserted — I replaced the backslashes with underscores in the call.
Then, the routing will be called but no mail is sent. I couldn’t dig further yet.

Am 11.04.2014 um 12:41 schrieb Brian J. Murrell notifications@github.com:

**Comment by georgehrke** [comment]
> just compare the content of the patch with what I inserted — I replaced the backslashes with underscores in the call.

Did you test on stable6 branch?

**Comment by joergmschulz** [comment]
yes, 6.0.2
Am 11.04.2014 um 14:27 schrieb Georg Ehrke notifications@github.com:

**Comment by brianjmurrell** [comment]
OK.  After a bit of debugging I have figured out that the right code to at least set the IMipHandler is in fact:

``` php
        $caldavPlugin = new Sabre_CalDAV_Plugin();
        $caldavPlugin->setIMipHandler(
                new Sabre_CalDAV_Schedule_IMip($email)
        );
        $server->addPlugin($caldavPlugin); 
```

And doing that does call into _3rdparty/Sabre/CalDAV/Plugin.php:setIMipHandler()_

The function that actually is supposed to do the mailing is _3rdparty/Sabre/CalDAV/Schedule/IMip.php:sendMessage()_ and the only caller of that is _Sabre_CalDAV_Plugin:iMIPMessage()_ which calls down through this stack:

_unknownMethod()_
_outboxRequest()_
_handleEventNotification()_
_iMIPMessage()_

_unknownMethod()_ is registered in _class Sabre_CalDAV_Plugin:initialize()_:

``` php
        $server->subscribeEvent('unknownMethod',array($this,'unknownMethod'));
```

The problem seems to be that _Sabre_CalDAV_Plugin:unknownMethod()_ is never called through the process of submitting a meeting request.  Thus, no meeting request e-mails.

**Comment by brianjmurrell** [comment]
Looking at SabreDav's current master vs. what is in owncloud, it seems there has been some significant development.  I suppose it's entirely possible that this functionality was not working at the version of SabreDav that's in owncloud.

I wonder how much API change there has been and whether getting a more up-to-date SabreDav to drop in is feasible.  I fear not.

**Comment by joergmschulz** [comment]
6.03 does not change the situation.
Is there still hope that someone familiar with PHP and the calendar/SabreDAV architecture progresses this one?

File: appinfo/remote.php [hunk 1]
```diff
@@ -43,6 +43,14 @@
 $server->addPlugin(new Sabre_DAVACL_Plugin());
 $server->addPlugin(new Sabre_DAV_Browser_Plugin(false)); // Show something in the Browser, but no upload
 $server->addPlugin(new Sabre_CalDAV_ICSExportPlugin());
+$email = \OCP\Config::getUserValue(\OCP\User::getUser(), 'settings', 'email');
+if($email !== null) {
+	$caldavPlugin = new \Sabre\CalDAV\Plugin();
```
Comment (libasys) [inline appinfo/remote.php:51]: this code can't work, cause you are working here with namespace OOP but in the base.php only Classes with "Sabre_" gets autoloaded, so there occurs an error!
correct definition:

```
if($email !== null) {
    $caldavPlugin = new Sabre_CalDAV_Plugin();
    $caldavPlugin->setIMipHandler(
        new Sabre_CalDAV_Schedule_IMip($email)
    );
    $server->addPlugin($caldavPlugin);
}
```

**Comment by brianjmurrell** [comment]
@libasys: indeed.  I think that was already discussed in this thread above.

But that is the least of the problems.  Ultimately the problem is that this whole `setIMipHandler()` approach will not achieve what I, and I suspect most people here are hoping it will and that's to send out meeting invitations for all meetings scheduled by any/all clients.

`setIMipHandler()` only covers the very specific use-case of sending out calendar invites when the client doing the scheduling is a/the broken iCal (the Apple calendar client AFAIU) client.  Setting up this handler will have no effect for any other clients.

This problem/feature needs to be handled more generically, per https://groups.google.com/forum/#!topic/sabredav-discuss/yyeJSYKs8SQ

**Comment by libasys** [comment]
@brianjmurrell my fault, i didn't read the comments on this pull request. i only saw the pr :)!
The discussion on google group sounds interesting, is there an update about the development of the shedule feature? about imiphandler: oc have to update the 3rdparty to sabredav version 1.8.xxx, but that's a lot of work for the devs affects also the dev on core, cause all sabreDav stuff needs to update to namespace php, and new abstract backends implementation! every app which useses caldav,webdav or carddav need this new code definition. lib files (e.g. /connector) also. So I think this comes with version newer than oc 7 or 8!

**Comment by libasys** [comment]
Opend new Issue on the core for missing email address in the principal scheme:
https://github.com/owncloud/core/issues/8515

**Comment by ghost** [comment]
Hi,

not directly related to this but as the main issues about invites is locked i want to post this in here:

Starting with 2.1 SabreDAV is fully supporting event schedules and invitations:

http://sabre.io/dav/scheduling/

Ref: https://github.com/owncloud/core/issues/12876 to update SabreDAV in core

**Comment by DeepDiver1975** [comment]
https://github.com/owncloud/calendar/pull/881 was merged

**Pull request closed without merging**